android/walletlib/src/main/java/com/solana/mobilewalletadapter/walletlib/scenario/LocalScenario.java

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright (c) 2022 Solana Mobile Inc.
+ */
+
+package com.solana.mobilewalletadapter.walletlib.scenario;
+
+import android.content.Context;
+import android.net.Uri;
+import android.os.Handler;
+import android.os.Looper;
+import android.util.Log;
+
+import androidx.annotation.GuardedBy;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.solana.mobilewalletadapter.common.protocol.MessageReceiver;
+import com.solana.mobilewalletadapter.common.protocol.MobileWalletAdapterSessionCommon;
+import com.solana.mobilewalletadapter.common.util.NotifyingCompletableFuture;
+import com.solana.mobilewalletadapter.walletlib.authorization.AuthRecord;
+import com.solana.mobilewalletadapter.walletlib.authorization.AuthRepository;
+import com.solana.mobilewalletadapter.walletlib.authorization.AuthRepositoryImpl;
+import com.solana.mobilewalletadapter.walletlib.authorization.AuthIssuerConfig;
+import com.solana.mobilewalletadapter.walletlib.protocol.MobileWalletAdapterConfig;
+import com.solana.mobilewalletadapter.walletlib.protocol.MobileWalletAdapterServer;
+import com.solana.mobilewalletadapter.walletlib.protocol.MobileWalletAdapterSession;
+import com.solana.mobilewalletadapter.walletlib.util.LooperThread;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public abstract class LocalScenario implements Scenario {
+    private static final String TAG = LocalScenario.class.getSimpleName();
+
+    @NonNull
+    final public byte[] associationPublicKey;
+
+    @NonNull
+    protected final MobileWalletAdapterConfig mMobileWalletAdapterConfig;
+    @NonNull
+    protected final Looper mIoLooper;
+    @NonNull
+    protected final Handler mIoHandler;
+    @NonNull
+    protected final Scenario.Callbacks mCallbacks;
+    @NonNull
+    protected final AuthRepository mAuthRepository;
+
+    private final Object mLock = new Object();
+    @Nullable
+    @GuardedBy("mLock")
+    private AuthRecord mActiveAuthorization = null;
+
+    private final PowerConfigProvider mPowerManager;
+
+    @Nullable
+    private ScheduledFuture<?> mNoConnectionTimeoutHandler;
+    private final ScheduledExecutorService mTimeoutExecutorService =
+            Executors.newSingleThreadScheduledExecutor();
+
+    protected LocalScenario(@NonNull Context context,
+                            @NonNull MobileWalletAdapterConfig mobileWalletAdapterConfig,
+                            @NonNull AuthIssuerConfig authIssuerConfig,
+                            @NonNull Scenario.Callbacks callbacks,
```

Comment:
WDYT about `LocalAssociationUri.createScenario` taking a `LocalScenario.Callbacks` instead of `Scenario.Callbacks`? That would allow this to be specifically typed, and we wouldn't need the type check on line 132.

Reply:
I tried that actually but because this overrides the `createScenario` method from `AssociationUri` I cant put a subtype of `Scenerio.Callbacks` there.

We could add an overload for `createScenario` to `AssociationUri` that takes `LocalScenario.callbacks` but I think that's not the best approach. Do you know of another way to resolve this?